### PR TITLE
Add code that can manage `PATH` by editing rc files

### LINKED ISSUE #1719: Write code to manage `PATH` by changing it in users' rc files
### Problem Description

Based on https://github.com/rancher-sandbox/rancher-desktop/discussions/1512, we should look at post-flight scripts to help set profiles to manage our paths for our tools, which are a mix of symlinks.

Python has an example we can leverage for the various shells available: https://github.com/python/cpython/blob/f4c03484da59049eb62a9bf7777b963e2267d187/Mac/BuildScript/scripts/postflight.patch-profile

### Proposed Solution

A similar script for Rancher Desktop

### PR DESCRIPTION
**Pull request by adamkpickering**
Closes #1719.

In the spirit of small PRs (if you can call this PR small), here is some of the work that I've done on changing `PATH` by altering users' rc files. It does not change anything with the way RD works currently. All this does is add code that will be used in a future PR, once the current set of code used for integrations is ready to be replaced.

### CODE REVIEW

File: src/integrations/__tests__/manageLinesInFile.spec.ts
Comment (mook-as) [inline src/integrations/__tests__/manageLinesInFile.spec.ts:96]: This test should start without the extra blank line, and ensure we do the right thing.

File: src/integrations/pathManager.ts
Comment (mook-as) [inline src/integrations/pathManager.ts:56]: Wouldn't this create a `~/.zshrc` even if I don't use zsh?

I'm not sure if that's good or not… Probably fine.

File: src/integrations/pathManager.ts
Comment (mook-as) [inline src/integrations/pathManager.ts:61]: I haven't used `csh` in decades; how do they deal with spaces in paths?

File: src/integrations/pathManager.ts
Comment (mook-as) [inline src/integrations/pathManager.ts:51]: This doesn't appear to deal with spaces in `$PATH` correctly. (Both our path, as well as the original one.)

File: src/integrations/pathManager.ts
Comment (mook-as) [inline src/integrations/pathManager.ts:71]: There appears to be a [`fish_add_path`](https://fishshell.com/docs/current/cmds/fish_add_path.html); would that be useful here?

If not, we should use long form here (`--export` instead of `-x`).

I'm not certain, but there appears to be an `--append` switch to modify the value in-place instead.

Speaking of, this appears to _prepend_ our path, whereas we _append_ in the other forms (I think?). Is that intended?

File: src/integrations/pathManager.ts
Comment (mook-as) [inline src/integrations/pathManager.ts:80]: We don't need to do the OS check here, do we? `XDG_CONFIG_HOME` would just never be set on mac.

**Review by mook-as — CHANGES_REQUESTED** [review]

File: src/integrations/pathManager.ts
Comment (adamkpickering) [reply src/integrations/pathManager.ts:56]: Yes it does. As far as I can tell this is the best way to do it. A couple reasons:

- We could introduce some mechanism to detect whether the user has a shell installed, and only write the relevant rc file if it is there. However there are edge cases here, like what happens if a user installs a shell after RD runs, or what happens if they uninstall the shell. It's feasible to do, but I don't think it's worth it given that...

- `RcFilePathManager` is intended for the "beginner" use case. It needs to just work and there shouldn't be any hiccups. The fact that it is aimed at beginners also means that we aren't likely to annoy anybody by cluttering up their system with rc files for shells they don't use.

That's the way I was thinking about it. We could introduce more logic here, but this way of managing integrations was supposed to be simpler than our current way.

File: src/integrations/manageLinesInFile.ts
Comment (adamkpickering) [reply src/integrations/manageLinesInFile.ts:13]: Done.

File: src/integrations/manageLinesInFile.ts
Comment (adamkpickering) [reply src/integrations/manageLinesInFile.ts:22]: Yes I think that would be the case. Is that alright?

File: src/integrations/__tests__/manageLinesInFile.spec.ts
Comment (adamkpickering) [reply src/integrations/__tests__/manageLinesInFile.spec.ts:19]: I think that checking whether it's falsy is all we can do - I've introduced code to do that.

File: src/integrations/manageLinesInFile.ts
Comment (mook-as) [reply src/integrations/manageLinesInFile.ts:22]: As long as we decide it is, it should be fine?
I do know that some _really_ old compilers and things would miss the last line (because they were looking for the line feed), but that's like, 80s or earlier, I think.

File: src/integrations/__tests__/manageLinesInFile.spec.ts
Comment (adamkpickering) [reply src/integrations/__tests__/manageLinesInFile.spec.ts:41]: Done.

File: src/integrations/__tests__/manageLinesInFile.spec.ts
Comment (adamkpickering) [reply src/integrations/__tests__/manageLinesInFile.spec.ts:51]: I'm a little confused - is there something I should change here? Or were you just commenting?

File: src/integrations/pathManager.ts
Comment (adamkpickering) [reply src/integrations/pathManager.ts:61]: It seems that setting `path` is fundamentally flawed with respect to spaces:

```
% echo $path
/sbin /usr/sbin /bin /usr/bin /usr/local/bin /usr/local/sbin /opt /home/adam/bin /home/adam/.local/bin /usr/local/go/bin .
% set path=("$path" 'with spaces')
% echo $path
/sbin /usr/sbin /bin /usr/bin /usr/local/bin /usr/local/sbin /opt /home/adam/bin /home/adam/.local/bin /usr/local/go/bin . with spaces
% echo $PATH
/sbin /usr/sbin /bin /usr/bin /usr/local/bin /usr/local/sbin /opt /home/adam/bin /home/adam/.local/bin /usr/local/go/bin .:with spaces
% set path=("$path" 'another one')
% echo $path
/sbin /usr/sbin /bin /usr/bin /usr/local/bin /usr/local/sbin /opt /home/adam/bin /home/adam/.local/bin /usr/local/go/bin . with spaces another one
% echo $PATH
/sbin /usr/sbin /bin /usr/bin /usr/local/bin /usr/local/sbin /opt /home/adam/bin /home/adam/.local/bin /usr/local/go/bin . with spaces:another one
```

With this, each time `path` is set the previous `path` elements get turned into one big element of `path`. Not good. The following was enlightening:

```
% echo $path
/sbin /usr/sbin /bin /usr/bin /usr/local/bin /usr/local/sbin /opt /home/adam/bin /home/adam/.local/bin /usr/local/go/bin .
% echo $PATH
/sbin:/usr/sbin:/bin:/usr/bin:/usr/local/bin:/usr/local/sbin:/opt:/home/adam/bin:/home/adam/.local/bin:/usr/local/go/bin:.
% setenv PATH $PATH\:"with spaces"
% echo $path
/sbin /usr/sbin /bin /usr/bin /usr/local/bin /usr/local/sbin /opt /home/adam/bin /home/adam/.local/bin /usr/local/go/bin . with spaces
% echo $PATH
/sbin:/usr/sbin:/bin:/usr/bin:/usr/local/bin:/usr/local/sbin:/opt:/home/adam/bin:/home/adam/.local/bin:/usr/local/go/bin:.:with spaces
% setenv PATH $PATH\:"more spaces"
setenv: Too many arguments.
% setenv PATH "$PATH"\:"more spaces"
% echo $PATH
/sbin:/usr/sbin:/bin:/usr/bin:/usr/local/bin:/usr/local/sbin:/opt:/home/adam/bin:/home/adam/.local/bin:/usr/local/go/bin:.:with spaces:more spaces
% echo $path
/sbin /usr/sbin /bin /usr/bin /usr/local/bin /usr/local/sbin /opt /home/adam/bin /home/adam/.local/bin /usr/local/go/bin . with spaces more spaces
```

So, `setenv PATH "$PATH"\:"ourPath"` should work.

File: src/integrations/pathManager.ts
Comment (adamkpickering) [reply src/integrations/pathManager.ts:51]: Fixed.

File: src/integrations/pathManager.ts
Comment (adamkpickering) [reply src/integrations/pathManager.ts:80]: Right you are, fixed.

File: src/integrations/__tests__/manageLinesInFile.spec.ts
Comment (mook-as) [reply src/integrations/__tests__/manageLinesInFile.spec.ts:51]: No, just commenting, and double checking to make sure you're aware of this.

File: src/integrations/pathManager.ts
Comment (adamkpickering) [reply src/integrations/pathManager.ts:71]: Looks like `fish_add_path` [was added in fish 3.2.0](https://fishshell.com/docs/current/relnotes.html). I think that's too recent; for example I get version 3.1.0 when I install via `apt`. Otherwise it would be great.

> If not, we should use long form here (--export instead of -x).
>
> I'm not certain, but there appears to be an --append switch to modify the value in-place instead.

Done, we're using `--export` and `--append` now.

> Speaking of, this appears to prepend our path, whereas we append in the other forms (I think?). Is that intended?

No it wasn't, nice catch. I was working off of the python link in https://github.com/rancher-sandbox/rancher-desktop/issues/1719 and it looks like copy-paste got away on me :)

File: src/integrations/__tests__/manageLinesInFile.spec.ts
Comment (mook-as) [reply src/integrations/__tests__/manageLinesInFile.spec.ts:135]: Actually, should this be `await expect(…).rejects.not.toThrow()`?

**Review by mook-as — CHANGES_REQUESTED** [review]
Only thing that might block merging is the a last assertion (`expect()….rejects`).

File: src/integrations/pathManager.ts
Comment (jandubois) [inline src/integrations/pathManager.ts:51]: It looks like we are appending the integrations at the end of the `PATH`.

I was under the assumption that we would be prepending to the front, for the benefit of the inexperienced user who might have an older version of the utility already somewhere else in their `PATH`.

We offer the "do-it-yourself" option for users who want to precisely control shadowing of binaries on the `PATH`.

**Review by jandubois — CHANGES_REQUESTED** [review]

File: src/integrations/pathManager.ts
Comment (adamkpickering) [reply src/integrations/pathManager.ts:7]: That is definitely odd. Maybe we'll leave it for now? It's an easy fix if it does cause any problems

File: src/integrations/__tests__/manageLinesInFile.spec.ts
Comment (adamkpickering) [reply src/integrations/__tests__/manageLinesInFile.spec.ts:13]: According to https://playwright.dev/docs/test-parallel:

> By default, test files are run in parallel. Tests in a single file are run in order, in the same worker process.

It seems fine the way it is. But I'm still learning the unit testing side of RD so I could be wrong.

File: src/integrations/pathManager.ts
Comment (adamkpickering) [reply src/integrations/pathManager.ts]: That's fair. I've now put the integrations directory at at the head of the `PATH`.

File: src/integrations/__tests__/manageLinesInFile.spec.ts
Comment (mook-as) [reply src/integrations/__tests__/manageLinesInFile.spec.ts]: With a bunch of trial and error, this seems to work:
```ts
test('Do nothing when desiredPresent is false and file does not exist', async() => {
  await expect(manageLinesInFile(rcFilePath, [TEST_LINE_1], false)).resolves.not.toThrow();
});
```
I've also checked that throwing in `manageLinesInFile` in that case actually triggers a failure.  I don't think the `.not.toThrow()` there is actually doing anything though; I think by the point it calculates `.resolves` it already checks that it wasn't rejected.

In this case, we also actually do an `await`, so we happen to no longer need the eslint comments. Bonus!

Comment (adamkpickering) [reply src/integrations/__tests__/manageLinesInFile.spec.ts]: I forgot to say that I had removed the lines in my comment above.

Cool, I've changed the test to what you wrote above.

File: src/integrations/__tests__/manageLinesInFile.spec.ts
Comment (mook-as) [reply src/integrations/__tests__/manageLinesInFile.spec.ts:13]: Well, we seem to be able to use `testDir` correctly, somehow, so that's _probably_ okay?

File: src/integrations/pathManager.ts
Comment (mook-as) [inline src/integrations/pathManager.ts:89]: Hmm, currently `pathManager` is initially unset, isn't it? (I think TypeScript didn't catch this because there's some iffiness around initially unset things?)

File: src/integrations/pathManager.ts
Comment (mook-as) [reply src/integrations/pathManager.ts:7]: I'm okay with the type (it probably just figured it's the same shape, so whatever); however, `pathManager` is unset, so things would blow up when we call `setPathManagementStrategy` the first time.  We currently never do that, so things are okay-ish.  But let's fix it here so we can have a good place to start from?

**Review by mook-as — CHANGES_REQUESTED** [review]

**Review by mook-as — APPROVED** [review]

**Review by mook-as — APPROVED** [review]
GitHub appears to have the previous approval? (It also lost Jan's disapproval, so I guess it cancels out?)

**Review by jandubois — APPROVED** [review]
Thanks, LGTM! Sorry for the long delay in reviewing this!

**Pull request merged**
